src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.References.cs

```diff
@@ -123,43 +115,55 @@ internal static void AssertThreadAffinity(object value)
         /// <inheritdoc />
         public override string ToString() => $"(js-obj js '{JSHandle}')";
 
+        // when we know that JS side already freed the handle
         internal void DisposeLocal()
         {
-            JSHostImplementation.ThreadCsOwnedObjects.Remove(JSHandle);
-            _isDisposed = true;
-            JSHandle = IntPtr.Zero;
+#if FEATURE_WASM_THREADS
+            lock (_lockObject)
+#endif
+            {
+                ProxyContext.RemoveCSOwnedObject(JSHandle);
```

Comment:
i think this general pattern i see here (acquire the object lock, then call into the proxy context which acquires its own lock) is going to create lots of deadlocks, since another thread could first acquire the proxy context lock and then attempt to manipulate this JSObject. I don't think we can perform proxycontext calls while holding other locks.

Reply:
well, perhaps I don't need JSObject lock anymore!